File: backend/schemas/daily_task.py
from pydantic import BaseModel, Field
from datetime import date, time, datetime
from typing import Optional,List,Union
from backend.models.enums import CategoryType, RepeatType, PriorityType

class DailySubtaskCreate(BaseModel):
    text: str
    done: Optional[bool] = False

class DailyNoteCreate(BaseModel):
    text: str
class DailyTaskBase(BaseModel):
    """
    Günlük görev için temel alanları tanımlar.
    """
    title: str
    description: Optional[str] = None
    date: date
    start_time: Optional[time] = None
    duration_minutes: Optional[int] = None
    calendar_event_id: Optional[str] = None
    learning_goal_id: Optional[int] = None
    category: Optional[CategoryType] = CategoryType.other
    tags: List[str] = []
    repeat: Optional[RepeatType] = RepeatType.none
    custom_repeat: Optional[str] = None
    priority: Optional[PriorityType] = PriorityType.medium


class DailyTaskCreate(DailyTaskBase):
    user_id: int
    subtasks: Optional[List[DailySubtaskCreate]] = []
    notes: Optional[List[DailyNoteCreate]] = []

class DailyTaskUpdate(BaseModel):
    """
    Günlük görevi güncellemek için şema.
    """
    title: Optional[str] = None
    description: Optional[str] = None
    date: Optional[Union[date, str]] = None
    start_time: Optional[time] = None
    duration_minutes: Optional[int] = None
    is_completed: Optional[bool] = None
    calendar_event_id: Optional[str] = None
    learning_goal_id: Optional[int] = None
    category: Optional[CategoryType] = None
    tags: Optional[List[str]] = Field(default_factory=list)
    repeat: Optional[RepeatType] = None
    custom_repeat: Optional[str] = None
    priority: Optional[PriorityType] = None
    subtasks: Optional[List[DailySubtaskCreate]] = Field(default_factory=list)
    notes: Optional[List[DailyNoteCreate]] = Field(default_factory=list)

    model_config = {
        "from_attributes": True,
        "use_enum_values": True
    }
class DailyTaskOut(DailyTaskBase):
    id: int
    user_id: int
    is_completed: bool
    created_at: datetime
    updated_at: datetime
    subtasks: List[DailySubtaskCreate] = []
    notes: List[DailyNoteCreate] = []

    model_config = {
        "from_attributes": True,
        "use_enum_values": True

    }

File: frontend/src/types/dailyTask.ts
// Tekrar eden görev türleri
export type RepeatType = 'none' | 'daily' | 'weekly' | 'weekdays' | 'monthly' | 'yearly' | 'custom';

// Öncelik seviyeleri
export type PriorityType = 'low' | 'medium' | 'high';

// Kategoriler
export type CategoryType = 'study' | 'exam' | 'project' | 'homework' | 'lab' | 'language' | 'reading' | 'other';

// Alt görev tipi
export interface Subtask {
  text: string;
  done?: boolean;
}

// Not tipi
export interface Note {
  text: string;
}

// Yeni görev oluşturmak için gereken alanlar
export interface DailyTaskCreate {
  title: string;
  description?: string;
  date: string;           // "YYYY-MM-DD"
  start_time?: string;    // "HH:MM:SS" or ISO datetime
  duration_minutes?: number;
  learning_goal_id?: number;
  calendar_event_id?: string;
  category?: CategoryType;
  tags?: string[];
  repeat?: RepeatType;
  custom_repeat?: string;
  priority?: PriorityType;
  user_id: number;
  subtasks?: Subtask[];
  notes?: Note[];
}

// Görev güncelleme için tüm alanlar opsiyonel
export interface DailyTaskUpdate {
  title?: string;
  description?: string;
  date?: string;
  start_time?: string;
  duration_minutes?: number;
  is_completed?: boolean;
  learning_goal_id?: number;
  calendar_event_id?: string;
  category?: CategoryType;
  tags?: string[];
  repeat?: RepeatType;
  custom_repeat?: string;
  priority?: PriorityType;
  subtasks?: Subtask[];
  notes?: Note[];
}

// API'den dönen tam görev verisi
export interface DailyTaskOut {
  id: number;
  title: string;
  description?: string;
  date: string;
  start_time?: string;
  duration_minutes?: number;
  calendar_event_id?: string;
  learning_goal_id?: number;
  category?: string;
  tags: string[];
  repeat?: string;
  custom_repeat?: string;
  priority?: string;
  user_id: number;
  is_completed: boolean;
  created_at: string;
  updated_at: string;
  subtasks?: Subtask[];
  notes?: Note[];
}